File: tools/openings.py
"""Windows and doors, and what their existence proves about the walls.

A door is evidence a wall is there. That is the useful direction, and it is why
this module exists separately from the wall tracing rather than after it: no
amount of looking at a pair of parallel lines distinguishes a stud wall from a
countertop, because the drawing draws them the same, but nobody puts a door in
a countertop. Every opening found here anchors the line it sits on as real.

It also fixes a counting problem. Rooms merge when the wall between them is
missing, and the wall between two rooms is almost always the one with the door
in it -- that is how you get from one to the other. So finding the doors and
finding the rooms are the same job approached from opposite ends.

Three symbols, and none of them is a gap in the wall:

  swing    A leaf drawn at an angle plus its arc. The arc is a POLYLINE, not a
           bezier -- the only true curves on this sheet are the toilet, the tub
           and the sink -- so it is found by clustering short diagonal chords.

  slider   Two thin panels, each about 25in by 1.3in, OFFSET across the wall by
           an inch or two and OVERLAPPING along it. Closets get these, which is
           why closets never separated while only swings were detected.

  glazing  Line pairs about an inch apart lying ON the wall's centreline, side
           by side and NOT overlapping, one per sash.

The slider and the glazing are the pair worth being careful about: both are
thin parallel lines inside a wall, and reading a closet's bypass doors as a
window is an easy mistake. What separates them is that slider panels are offset
across the wall and overlap along it, while sashes share the centreline and sit
end to end.

Everything is then widened out to the jambs either side. Jambs are required,
not optional: they are what says the wall is actually interrupted, and without
them every cabinet door in the kitchen reads as a closet.
"""

PANEL_GAP_IN = (0.8, 2.6)      # thickness of a drawn door panel or sash
PANEL_MIN_IN = 16.0
SLIDER_OFFSET_IN = (1.2, 4.5)  # how far the two leaves sit apart across the wall
SLIDER_OVERLAP_IN = 6.0
GLAZING_GAP_IN = (0.4, 2.2)
GLAZING_MIN_IN = 6.0
GLAZING_JOIN_IN = 3.0
ON_WALL_IN = 2.5  # tight: 4.0 let cabinet lines read as glazing (15 windows, some 10ft)
JAMB_SPAN_IN = (3.0, 12.0)     # a jamb crosses the wall, so it is wall-thick
JAMB_PAIR_IN = 2.5
JAMB_STRADDLE_IN = 3.0
BRACKET_SLACK_IN = 8.0
MIN_OPENING_IN, MAX_OPENING_IN = 18.0, 144.0
SWING_SIDE_IN = (18.0, 44.0)
SWING_MAX_CHORDS = 14


def thin_pairs(runs, gap=PANEL_GAP_IN, min_length=PANEL_MIN_IN):
    """Thin rectangles drawn along a wall: door leaves and window sashes."""
    out = []
    runs = sorted(runs)
    for i, (c1, a1, b1) in enumerate(runs):
        for c2, a2, b2 in runs[i + 1:]:
            separation = c2 - c1
            if separation > gap[1]:
                break
            if separation < gap[0]:
                continue
            lo, hi = max(a1, a2), min(b1, b2)
            if hi - lo >= min_length:
                out.append(((c1 + c2) / 2.0, lo, hi))
    return out


def sliders(panels):
    """Leaf pairs that bypass: offset across the wall, overlapping along it."""
    found, seen = [], set()
    for i, (c1, a1, b1) in enumerate(panels):
        for c2, a2, b2 in panels[i + 1:]:
            offset = abs(c2 - c1)
            if not SLIDER_OFFSET_IN[0] <= offset <= SLIDER_OFFSET_IN[1]:
                continue
            if min(b1, b2) - max(a1, a2) < SLIDER_OVERLAP_IN:
                continue
            span = (min(a1, a2), max(b1, b2))
            key = (round(span[0]), round(span[1]), round((c1 + c2) / 2))
            if key not in seen:
                seen.add(key)
                found.append(((c1 + c2) / 2.0, span[0], span[1]))
    return found


def glazing(runs):
    """Sashes: thin pairs on the centreline, merged across their mullions."""
    from extract import merge_collinear
    pairs = thin_pairs(runs, GLAZING_GAP_IN, GLAZING_MIN_IN)
    return [tuple(run) for run in
            merge_collinear([list(p) for p in pairs],
                            tol=GLAZING_GAP_IN[1], join=GLAZING_JOIN_IN)]


def swings(diagonals):
    """Door swing arcs, clustered out of the short chords that draw them."""
    groups = []
    for box in diagonals:
        for group in groups:
            if not (box[0] > group[2] + 3 or box[2] < group[0] - 3
                    or box[1] > group[3] + 3 or box[3] < group[1] - 3):
                group[0], group[1] = min(group[0], box[0]), min(group[1], box[1])
                group[2], group[3] = max(group[2], box[2]), max(group[3], box[3])
                group[4] += 1
                break
        else:
            groups.append([box[0], box[1], box[2], box[3], 1])

    out = []
    for x0, y0, x1, y1, chords in groups:
        width, height = x1 - x0, y1 - y0
        # Square-ish, door-sized, and cheap to draw. The hatched cabinet fills
        # are hundreds of chords in a box this size and would read as one very
        # large door.
        if (chords <= SWING_MAX_CHORDS
                and SWING_SIDE_IN[0] <= width <= SWING_SIDE_IN[1]
                and SWING_SIDE_IN[0] <= height <= SWING_SIDE_IN[1]
                and 0.6 <= width / height <= 1.6):
            out.append((x0, y0, x1, y1))
    return out


def jambs_on(wall, perpendicular):
    """Where a wall is interrupted, in wall-length coordinates."""
    coord, lo, hi = wall
    marks = []
    for pc, pa, pb in perpendicular:
        if not JAMB_SPAN_IN[0] <= pb - pa <= JAMB_SPAN_IN[1]:
            continue
        # Contains the centreline within a tolerance rather than strictly
        # straddling it. Clustering the canonical axes moves a wall line by up
        # to the cluster tolerance, so a jamb drawn dead centre on the real wall
        # can sit wholly to one side of the line that represents it -- which
        # rejected every closet slider in the plan and found zero of them.
        if pa - JAMB_STRADDLE_IN <= coord <= pb + JAMB_STRADDLE_IN and lo - 2 <= pc <= hi + 2:
            marks.append(pc)
    grouped = []
    for mark in sorted(marks):
        if grouped and mark - grouped[-1][-1] <= JAMB_PAIR_IN:
            grouped[-1].append(mark)
        else:
            grouped.append([mark])
    return [sum(group) / len(group) for group in grouped]


def bracket(marks, a, b, slack=BRACKET_SLACK_IN):
    """Widen a span out to the jambs either side of it."""
    before = [m for m in marks if m <= a + slack]
    after = [m for m in marks if m >= b - slack]
    if not before or not after:
        return None
    lo, hi = max(before), min(after)
    return (lo, hi) if hi - lo >= MIN_OPENING_IN else None


def on_walls(walls, perpendicular, glazed, slid, swung, horizontal):
    """Match each symbol to the wall it sits in, and size it by the jambs."""
    out = []
    for coord, lo, hi in walls:
        marks = jambs_on((coord, lo, hi), perpendicular)

        for kind, items in (("window", glazed), ("slider", slid)):
            for centre, a, b in items:
                if abs(centre - coord) > ON_WALL_IN or a < lo - 12 or b > hi + 12:
                    continue
                span = bracket(marks, a, b)
                if span and MIN_OPENING_IN <= span[1] - span[0] <= MAX_OPENING_IN:
                    out.append((kind, (span[0] + span[1]) / 2.0, coord,
                                span[1] - span[0], horizontal))

        for x0, y0, x1, y1 in swung:
            near = (y0, y1) if horizontal else (x0, x1)
            along = (x0, x1) if horizontal else (y0, y1)
            if min(abs(near[0] - coord), abs(near[1] - coord)) > 6.0:
                continue
            if along[0] < lo - 12 or along[1] > hi + 12:
                continue
            span = bracket(marks, along[0], along[1]) or along
            if 20.0 <= span[1] - span[0] <= 60.0:
                out.append(("door", (span[0] + span[1]) / 2.0, coord,
                            span[1] - span[0], horizontal))
    return out


def dedupe(openings):
    """One opening per place. A slider outranks a window at the same spot."""
    rank = {"slider": 0, "door": 1, "window": 2}
    kept = []
    for kind, along, coord, width, horizontal in sorted(
            openings, key=lambda o: rank.get(o[0], 3)):
        x, y = (along, coord) if horizontal else (coord, along)
        if any(abs(x - kx) < 12 and abs(y - ky) < 12 for _, kx, ky, _, _ in kept):
            continue
        kept.append((kind, x, y, width, horizontal))
    return kept

File: app/src/app/import/openings.js
// @ts-check
/**
 * Windows and doors: where the wall stops, and what fills it.
 *
 * A port of `tools/opening_truth.py`, minus its scoring and rendering, which
 * are fixture tooling rather than pipeline.
 *
 * ## Two detectors, because they miss different things
 *
 * An opening is found twice over and the results are UNIONED, not intersected.
 *
 *   symbols   what the drafter drew in the wall: a swing arc off to one side,
 *             two overlapping bypass leaves, glazing running jamb to jamb.
 *             The only thing that finds a plain hinged doorway, whose swing is
 *             drawn out in the room so that nothing marks the wall line at all.
 *   gaps      where the wall's own face lines STOP. Structural rather than
 *             symbolic, so it catches a cased opening with nothing drawn in
 *             it, which no symbol reader can see.
 *
 * Requiring both to agree would find fewer than either alone.
 */

import {cover} from './walls.js';
import {pyRound} from './round.js';
import {
	ARCHITECTURE, SYMBOLS, diagonals, findOpenings, segments, swingBoxes,
} from './symbols.js';

/** Narrower than this is a break in the linework; wider is a room. */
const MIN_OPENING_IN = 12.0;
const MAX_OPENING_IN = 96.0;

/**
 * How close a face line must be to a box's own `near`/`far` to BE it.
 *
 * The one deliberate divergence from `tools/opening_truth.py`, which uses
 * 1e-6, and the reason is a precision bug in the reference rather than a
 * disagreement about the rule.
 *
 * `walls.coalesce` groups face pairs by their coordinates ROUNDED TO THREE
 * DECIMALS and then carries the rounded values forward as the box's `near` and
 * `far`. `find` then looks for the faces the box was built from by comparing
 * those rounded values against the RAW face coordinates - so the lookup only
 * succeeds when a coordinate happens to land within a millionth of an inch of
 * its own thousandth-rounding, which is luck rather than logic.
 *
 * Measured on the sample sheet, on the wall at y=308.117in:
 *
 *     box near, rounded    305.187000000000   both
 *     face coord, MuPDF    305.186859130859   1.4e-4 away - no match
 *     face coord, pdf.js   305.186850000000   1.4e-4 away - no match
 *     box far, rounded     311.047000000000   both
 *     face coord, MuPDF    311.046997070312   2.9e-6 away - no match
 *     face coord, pdf.js   311.047000284831   2.8e-7 away - MATCHES
 *
 * The two readers differ because MuPDF holds path coordinates as float32 and
 * pdf.js keeps the float64 it parsed. Neither is wrong; the comparison is. And
 * when it fails it fails SILENTLY and completely: `find` returns no gaps for
 * that wall, so the structural half of the opening detection - the half that
 * finds a cased opening with nothing drawn in it - is simply dead, and the
 * trace still succeeds.
 *
 * A thousandth of an inch is below anything a drawing distinguishes and above
 * the rounding the pipeline itself applied, so the lookup now means what it
 * says in both implementations. `tools/` still has the 1e-6 and should be
 * brought into line; it is left alone here because changing it changes what
 * the Python tracer finds on the house it is tuned against.
 */
const FACE_MATCH_IN = 1e-3;
/** How far an opening may sit from its counterpart and still be the same one. */
const PLACE_TOL_IN = 6.0;
/**
 * How far off a wall's centreline a swing box may sit and still belong to it.
 * A swing reaches a door's width into the room, so its box centre is about
 * half a door away from the wall.
 */
const SWING_ACROSS_IN = 48.0;

/**
 * @typedef {object} Opening
 * @property {string} kind window | door | cased | unknown
 * @property {boolean} horizontal
 * @property {number} centre The wall's centreline coordinate.
 * @property {number} thickness
 * @property {number} lo Along the wall.
 * @property {number} hi
 * @property {number} width_in
 * @property {string} [from]
 * @property {?string} [hinge]
 * @property {?string} [swing]
 * @property {boolean} [exterior]
 */

/**
 * Every gap in a wall's own face coverage, along the wall it belongs to.
 *
 * ONLY the pair the box was built from is read. Every other line in the band -
 * the window frame above all, which spans exactly the opening - fills the gap
 * that marks it.
 *
 * @param {ReturnType<import('./walls.js').trace>} traced
 * @returns {Opening[]}
 */
export function find(traced)
{
	const out = [];
	for (const box of traced.boxes)
	{
		const runs = traced.faces[box.horizontal ? 0 : 1];
		const mine = runs.filter(([coord]) =>
			Math.abs(coord - box.near) < FACE_MATCH_IN
			|| Math.abs(coord - box.far) < FACE_MATCH_IN);
		if (!mine.length)
		{
			continue;
		}
		const clipped = mine
			.map(([, lo, hi]) => [Math.max(lo, box.drawn_lo), Math.min(hi, box.drawn_hi)])
			.filter(([lo, hi]) => hi > lo);
		const covered = cover(clipped);
		for (let index = 1; index < covered.length; index += 1)
		{
			const width = covered[index][0] - covered[index - 1][1];
			if (!(width >= MIN_OPENING_IN && width <= MAX_OPENING_IN))
			{
				continue;
			}
			out.push({
				kind: 'unknown',
				horizontal: box.horizontal,
				centre: pyRound(box.centre, 2),
				thickness: pyRound(box.thickness, 2),
				lo: pyRound(covered[index - 1][1], 2),
				hi: pyRound(covered[index][0], 2),
				width_in: pyRound(width, 2),
			});
		}
	}
	return sortOpenings(out);
}

/**
 * What the drafter DREW in the wall: swing arcs, bypass leaves, glazing.
 *
 * Driven against the BOX walls rather than a lattice, so an opening lands on
 * the wall that will actually be written out instead of a few inches off it.
 *
 * @param {ReturnType<import('./walls.js').trace>} traced
 * @param {import('./readers/model.js').DrawnPath[]} paths
 * @param {number[]} clip
 * @param {number} scale
 * @param {{horizontal: Array<number[]>, vertical: Array<number[]>}} architecture
 * The structure layer's runs - passed in rather than re-derived, since the
 * caller has already chosen the layer and choosing it twice could differ.
 * @returns {Opening[]}
 */
export function findBySymbol(traced, paths, clip, scale, architecture)
{
	// Thickness goes with each wall so jambsOn can tell a jamb, which spans the
	// wall face to face, from a symbol drawn inside it.
	const wallsH = traced.boxes.filter((box) => box.horizontal)
		.map((box) => [box.centre, box.drawn_lo, box.drawn_hi, box.thickness]);
	const wallsV = traced.boxes.filter((box) => !box.horizontal)
		.map((box) => [box.centre, box.drawn_lo, box.drawn_hi, box.thickness]);

	const symbol = segments(paths, clip, scale, SYMBOLS);
	const swings = swingBoxes(diagonals(paths, clip, scale, SYMBOLS));

	const raw = [
		...findOpenings(wallsH, symbol.vertical, symbol.horizontal,
			architecture.horizontal, swings, true),
		...findOpenings(wallsV, symbol.horizontal, symbol.vertical,
			architecture.vertical, swings, false),
	];

	/** @type {Map<string, number>} */
	const thickness = new Map();
	for (const box of traced.boxes)
	{
		thickness.set(`${box.horizontal},${pyRound(box.centre, 1)}`, box.thickness);
	}

	return raw.map(([kind, along, coord, width, horizontal]) => ({
		kind,
		horizontal,
		centre: pyRound(coord, 2),
		thickness: pyRound(
			thickness.get(`${horizontal},${pyRound(coord, 1)}`) ?? 5.0, 2),
		lo: pyRound(along - width / 2, 2),
		hi: pyRound(along + width / 2, 2),
		width_in: pyRound(width, 2),
		from: 'symbol',
	}));
}

/**
 * One opening per place. First detector to claim a spot keeps it.
 *
 * @param {...Opening[]} groups
 * @returns {Opening[]}
 */
export function merge(...groups)
{
	/** @type {Opening[]} */
	const kept = [];
	for (const group of groups)
	{
		for (const opening of group)
		{
			let duplicate = false;
			for (const other of kept)
			{
				if (other.horizontal !== opening.horizontal)
				{
					continue;
				}
				if (Math.abs(other.centre - opening.centre) > PLACE_TOL_IN * 2)
				{
					continue;
				}
				if (Math.min(other.hi, opening.hi) - Math.max(other.lo, opening.lo) > 0)
				{
					duplicate = true;
					// A kind from the symbol pen beats "unknown" from a gap.
					if (other.kind === 'unknown' && opening.kind !== 'unknown')
					{
						other.kind = opening.kind;
					}
					break;
				}
			}
			if (!duplicate)
			{
				kept.push({...opening});
			}
		}
	}
	return sortOpenings(kept);
}

/**
 * Ask the symbol pen what is in each opening. Guesses, never invented.
 *
 * Only the classification is taken from symbols - the opening was already
 * located by the linework, so a symbol that cannot be read costs a label and
 * not a window.
 *
 * @param {Opening[]} openings
 * @param {import('./readers/model.js').DrawnPath[]} paths
 * @param {number[]} clip
 * @param {number} scale
 */
export function label(openings, paths, clip, scale)
{
	const symbol = segments(paths, clip, scale, SYMBOLS);
	const swings = swingBoxes(diagonals(paths, clip, scale, SYMBOLS));

	for (const opening of openings)
	{
		const along = [opening.lo, opening.hi];
		const centre = opening.centre;
		const near = swings.filter((swing) => within(swing, along, centre, opening.horizontal));
		if (near.length)
		{
			opening.kind = 'door';
			continue;
		}
		const runs = opening.horizontal ? symbol.horizontal : symbol.vertical;
		let glazing = 0;
		for (const [coord, lo, hi] of runs)
		{
			if (Math.abs(coord - centre) <= opening.thickness && hi > along[0] && lo < along[1])
			{
				glazing += Math.min(hi, along[1]) - Math.max(lo, along[0]);
			}
		}
		const span = along[1] - along[0];
		if (span > 0 && glazing / span >= 0.6)
		{
			opening.kind = 'window';
		}
		else if (glazing <= 1.0)
		{
			opening.kind = 'cased';
		}
	}
	return openings;
}

/**
 * @param {number[]} swing
 * @param {number[]} along
 * @param {number} centre
 * @param {boolean} horizontal
 */
function within(swing, along, centre, horizontal)
{
	const [x0, y0, x1, y1] = swing;
	const lo = horizontal ? x0 : y0;
	const hi = horizontal ? x1 : y1;
	const across = horizontal ? (y0 + y1) / 2 : (x0 + x1) / 2;
	return (hi > along[0] - PLACE_TOL_IN && lo < along[1] + PLACE_TOL_IN
		&& Math.abs(across - centre) <= 60.0);
}

/**
 * Which end a door is hinged on, and which way it swings.
 *
 * Both come off the swing arc's bounding box, and between them they are the
 * door's handing. The box is a square whose side is the door's width with one
 * corner at the hinge: the corner nearest a jamb says which END the hinge is
 * on, and which side of the wall the box sits on says which way it opens.
 * Four combinations, which is exactly the four handings a joiner would name.
 *
 * Matched one to one, nearest first. A loose match let a single swing be
 * claimed by three different doors at once.
 *
 * A door with no swing keeps hinge null - not a failure: a bypass slider does
 * not swing, and neither does a cased opening.
 *
 * @param {Opening[]} openings
 * @param {import('./readers/model.js').DrawnPath[]} paths
 * @param {number[]} clip
 * @param {number} scale
 */
export function hand(openings, paths, clip, scale)
{
	const swings = swingBoxes(diagonals(paths, clip, scale, SYMBOLS));
	const doors = openings.filter((opening) => opening.kind === 'door');
	for (const door of doors)
	{
		door.hinge = null;
		door.swing = null;
	}

	const pairs = [];
	for (const door of doors)
	{
		const centre = (door.lo + door.hi) / 2;
		swings.forEach((box, index) =>
		{
			const [x0, y0, x1, y1] = box;
			const along = door.horizontal ? (x0 + x1) / 2 : (y0 + y1) / 2;
			const across = door.horizontal ? (y0 + y1) / 2 : (x0 + x1) / 2;
			if (Math.abs(across - door.centre) > SWING_ACROSS_IN)
			{
				return;
			}
			pairs.push({gap: Math.abs(along - centre), door, index, box});
		});
	}

	const taken = new Set();
	for (const {gap, door, index, box} of pairs.sort((one, two) => one.gap - two.gap))
	{
		if (taken.has(index) || door.hinge !== null)
		{
			continue;
		}
		if (gap > (door.hi - door.lo))
		{
			continue;
		}
		taken.add(index);
		const [x0, y0, x1, y1] = box;
		const lo = door.horizontal ? x0 : y0;
		const hi = door.horizontal ? x1 : y1;
		const across = door.horizontal ? (y0 + y1) / 2 : (x0 + x1) / 2;
		door.hinge = Math.abs(lo - door.lo) < Math.abs(hi - door.hi) ? 'lo' : 'hi';
		door.swing = across > door.centre ? 'positive' : 'negative';
	}
	return openings;
}

/**
 * Python's `sorted(key=lambda o: (not o["horizontal"], o["centre"], o["lo"]))`
 * - horizontal walls first, then by position.
 *
 * @param {Opening[]} openings
 */
function sortOpenings(openings)
{
	return openings.slice().sort((one, two) =>
		(Number(!one.horizontal) - Number(!two.horizontal))
		|| (one.centre - two.centre)
		|| (one.lo - two.lo));
}

export {ARCHITECTURE, MAX_OPENING_IN, MIN_OPENING_IN};
